File: client/internal/publisher/publisher.go
package publisher

import (
	"context"
	"errors"
	"log"
	"os/exec"
	"strings"
	"sync"
	"syscall"
	"time"

	"robot-media-client/internal/config"
	"robot-media-client/internal/model"
)

type Publisher interface {
	Start(ctx context.Context, command model.StartCommand, rtspURL string) (string, string, error)
	Stop(sessionID string) error
	StopAll() error
}

// ProcessPublisher 用外部进程完成 RTSP -> LiveKit 的发布。
// 默认走 gst-launch/livekit-gstreamer-publisher，也可以通过 PUBLISHER_CMD 注入自定义命令。
type ProcessPublisher struct {
	cfg                    *config.Config
	cmds                   map[string]*processEntry
	sessions               map[string]string
	streamSessions         map[string]map[string]struct{}
	gstreamerFailedRTSPURL map[string]time.Time
	unexpectedExits        uint64
	tokenExpirations       uint64
	forcedKills            uint64
	lastCleanupMillis      int64
	maxCleanupMillis       int64
	events                 chan LifecycleEvent
	mu                     sync.Mutex
}

type processEntry struct {
	cmd       *exec.Cmd
	done      chan processResult
	mode      string
	expiresAt time.Time
}

type processResult struct {
	err      error
	exitedAt time.Time
}

type Snapshot struct {
	ActivePublishers  int    `json:"activePublishers"`
	ActiveSessions    int    `json:"activeSessions"`
	UnexpectedExits   uint64 `json:"unexpectedExits"`
	TokenExpirations  uint64 `json:"tokenExpirations"`
	ForcedKills       uint64 `json:"forcedKills"`
	LastCleanupMillis int64  `json:"lastCleanupMillis"`
	MaxCleanupMillis  int64  `json:"maxCleanupMillis"`
}

type LifecycleEvent struct {
	SessionIDs []string
	ReasonCode string
	Message    string
}

func NewProcessPublisher(cfg config.Config) *ProcessPublisher {
	return &ProcessPublisher{
		cfg:                    &cfg,
		cmds:                   make(map[string]*processEntry),
		sessions:               make(map[string]string),
		streamSessions:         make(map[string]map[string]struct{}),
		gstreamerFailedRTSPURL: make(map[string]time.Time),
		events:                 make(chan LifecycleEvent, 64),
	}
}

func (p *ProcessPublisher) Start(ctx context.Context, command model.StartCommand, rtspURL string) (string, string, error) {
	if !tokenUsable(command.ExpiresAt) {
		return "", "", errors.New("发布 Token 缺失、已过期或剩余有效期不足 30 秒")
	}
	p.mu.Lock()
	defer p.mu.Unlock()
	key := streamKey(command)
	previousKey := p.unbindSessionLocked(command.SessionID)
	if previousKey != "" && previousKey != key && len(p.streamSessions[previousKey]) == 0 {
		_ = p.stopStreamLocked(previousKey)
	}
	if entry := p.cmds[key]; entry != nil {
		if entryRunning(entry) && tokenUsable(entry.expiresAt) && tokenUsable(command.ExpiresAt) {
			p.bindSessionLocked(command.SessionID, key)
			return "TR_" + command.SessionID, trackName(command), nil
		}
		_ = p.stopStreamLocked(key)
	}
	trackName := "video." + command.Channel + "." + command.Quality
	if p.cfg.PublisherCmd != "" {
		return p.startCommand(ctx, command, rtspURL, trackName, key, p.cfg.PublisherCmd, "custom")
	}
	if p.cfg.PublisherMode == "ffmpeg" {
		if p.cfg.FFmpegPublisherCmd == "" {
			return "", "", errors.New("PUBLISHER_MODE=ffmpeg requires FFMPEG_PUBLISHER_CMD")
		}
		return p.startCommand(ctx, command, rtspURL, trackName, key, p.cfg.FFmpegPublisherCmd, "ffmpeg")
	}
	if p.cfg.PublisherMode != "auto" && p.cfg.PublisherMode != "gstreamer" {
		return "", "", errors.New("unsupported PUBLISHER_MODE: " + p.cfg.PublisherMode)
	}
	if p.shouldStartWithFFmpeg(command, rtspURL) {
		return p.startCommand(ctx, command, rtspURL, trackName, key, p.cfg.FFmpegPublisherCmd, "ffmpeg-first")
	}
	trackSid, publishedTrackName, err := p.startGStreamer(ctx, command, rtspURL, trackName, key)
	if err == nil {
		if p.cfg.PublisherMode == "auto" && p.cfg.FFmpegPublisherCmd != "" {
			p.watchGStreamerForFallback(ctx, command, rtspURL, trackName, key)
		} else if entry := p.cmds[key]; entry != nil {
			p.watchProcessExit(key, entry, command.SessionID)
		}
		return trackSid, publishedTrackName, nil
	}
	if p.cfg.PublisherMode == "gstreamer" || p.cfg.FFmpegPublisherCmd == "" {
		return trackSid, publishedTrackName, err
	}
	p.gstreamerFailedRTSPURL[rtspURL] = time.Now()
	log.Printf("GStreamer 推流失败，回退到 FFmpeg，会话ID=%s：%v", command.SessionID, err)
	return p.startCommand(ctx, command, rtspURL, trackName, key, p.cfg.FFmpegPublisherCmd, "ffmpeg")
}

func (p *ProcessPublisher) startCommand(ctx context.Context, command model.StartCommand, rtspURL string, trackName string, key string, template string, mode string) (string, string, error) {
	args := strings.Fields(template)
	if len(args) == 0 {
		return "", "", errors.New("publisher command is empty")
	}
	for i := range args {
		args[i] = strings.NewReplacer(
			"{rtsp}", rtspURL,
			"{livekitUrl}", command.LiveKitURL,
			"{token}", command.PublisherToken,
			"{room}", command.RoomName,
			"{track}", trackName,
		).Replace(args[i])
	}
	cmd := exec.CommandContext(ctx, args[0], args[1:]...)
	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
	if err := cmd.Start(); err != nil {
		return "", "", err
	}
	// 外部进程参数和输出可能包含 RTSP 凭据或 LiveKit Token，不写入应用日志。
	log.Printf("推流进程已启动，模式=%s 会话ID=%s", mode, command.SessionID)
	entry := newProcessEntry(cmd, mode, command.ExpiresAt)
	p.cmds[key] = entry
	p.bindSessionLocked(command.SessionID, key)
	if err := p.ensureRunning(key, entry); err != nil {
		return "", "", err
	}
	p.watchProcessExit(key, entry, command.SessionID)
	p.watchTokenExpiry(key, entry, command.SessionID)
	return "TR_" + command.SessionID, trackName, nil
}

func (p *ProcessPublisher) startGStreamer(ctx context.Context, command model.StartCommand, rtspURL string, trackName string, key string) (string, string, error) {
	// 默认 pipeline 只描述 GStreamer 的媒体处理部分；LiveKit URL/token 由 publisher 工具参数提供。
	pipeline := strings.NewReplacer(
		"{rtsp}", rtspURL,
		"{livekitUrl}", command.LiveKitURL,
		"{token}", command.PublisherToken,
		"{room}", command.RoomName,
		"{track}", trackName,
	).Replace(p.cfg.GStreamerPipeline)
	args := []string{"--url", command.LiveKitURL, "--token", command.PublisherToken, "--"}
	args = append(args, strings.Fields(pipeline)...)
	cmd := exec.CommandContext(ctx, p.cfg.GStreamerPublisherPath, args...)
	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
	if err := cmd.Start(); err != nil {
		return "", "", err
	}
	// 外部进程参数和输出可能包含 RTSP 凭据或 LiveKit Token，不写入应用日志。
	log.Printf("GStreamer 推流进程已启动，会话ID=%s", command.SessionID)
	entry := newProcessEntry(cmd, "gstreamer", command.ExpiresAt)
	p.cmds[key] = entry
	p.bindSessionLocked(command.SessionID, key)
	if err := p.ensureRunning(key, entry); err != nil {
		return "", "", err
	}
	p.watchTokenExpiry(key, entry, command.SessionID)
	return "TR_" + command.SessionID, trackName, nil
}

func (p *ProcessPublisher) shouldStartWithFFmpeg(command model.StartCommand, rtspURL string) bool {
	if p.cfg.PublisherMode != "auto" || p.cfg.FFmpegPublisherCmd == "" {
		return false
	}
	if failedAt, failed := p.gstreamerFailedRTSPURL[rtspURL]; failed {
		if time.Since(failedAt) < p.cfg.PublisherGStreamerRetry {
			return true
		}
		delete(p.gstreamerFailedRTSPURL, rtspURL)
		log.Printf("FFmpeg 回退冷却结束，重新尝试 GStreamer，会话ID=%s", command.SessionID)
	}
	return p.cfg.PublisherFFmpegFirstIDs[command.DeviceID]
}

func newProcessEntry(cmd *exec.Cmd, mode string, expiresAt time.Time) *processEntry {
	entry := &processEntry{cmd: cmd, done: make(chan processResult, 1), mode: mode, expiresAt: expiresAt}
	go func() {
		entry.done <- processResult{err: cmd.Wait(), exitedAt: time.Now()}
	}()
	return entry
}

func (p *ProcessPublisher) watchGStreamerForFallback(ctx context.Context, command model.StartCommand, rtspURL string, trackName string, key string) {
	entry := p.cmds[key]
	if entry == nil {
		return
	}
	go p.fallbackIfGStreamerExits(ctx, command, rtspURL, trackName, key, entry)
}

func (p *ProcessPublisher) fallbackIfGStreamerExits(ctx context.Context, command model.StartCommand, rtspURL string, trackName string, key string, entry *processEntry) {
	select {
	case result := <-entry.done:
		p.mu.Lock()
		defer p.mu.Unlock()
		if p.cmds[key] != entry {
			return
		}
		sessionIDs := p.sessionIDsLocked(key)
		delete(p.cmds, key)
		p.gstreamerFailedRTSPURL[rtspURL] = time.Now()
		p.recordCleanupLocked(result.exitedAt)
		p.unexpectedExits++
		log.Printf("GStreamer 推流进程异常退出，自动回退到 FFmpeg，会话ID=%s：%v", command.SessionID, result.err)
		if _, _, startErr := p.startCommand(ctx, command, rtspURL, trackName, key, p.cfg.FFmpegPublisherCmd, "ffmpeg"); startErr != nil {
			log.Printf("自动回退到 FFmpeg 失败，会话ID=%s：%v", command.SessionID, startErr)
			p.unbindStreamLocked(key)
			p.emitLocked(LifecycleEvent{SessionIDs: sessionIDs, ReasonCode: "PUBLISH_PROCESS_EXITED", Message: "推流进程异常退出且回退失败"})
		}
	case <-time.After(p.cfg.PublisherFallbackWatch):
		p.watchProcessExit(key, entry, command.SessionID)
	case <-ctx.Done():
		p.watchProcessExit(key, entry, command.SessionID)
	}
}

func (p *ProcessPublisher) Stop(sessionID string) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	key := p.sessions[sessionID]
	p.unbindSessionLocked(sessionID)
	if key == "" || len(p.streamSessions[key]) > 0 {
		return nil
	}
	return p.stopStreamLocked(key)
}

func (p *ProcessPublisher) StopAll() error {
	p.mu.Lock()
	defer p.mu.Unlock()
	// map 在遍历时会被 stopLocked 删除。Go 允许删除当前 map key，这里用于快速清空。
	for key := range p.cmds {
		_ = p.stopStreamLocked(key)
	}
	p.sessions = make(map[string]string)
	p.streamSessions = make(map[string]map[string]struct{})
	return nil
}

func (p *ProcessPublisher) stopStreamLocked(key string) error {
	entry := p.cmds[key]
	var err error
	if entry != nil && entry.cmd != nil && entry.cmd.Process != nil {
		err = syscall.Kill(-entry.cmd.Process.Pid, syscall.SIGTERM)
		if err != nil {
			err = entry.cmd.Process.Signal(syscall.SIGTERM)
		}
		p.forceKillAfterTimeout(entry)
	}
	delete(p.cmds, key)
	p.unbindStreamLocked(key)
	return err
}

func (p *ProcessPublisher) ensureRunning(key string, entry *processEntry) error {
	select {
	case result := <-entry.done:
		// 进程两秒内退出通常表示 pipeline 参数、RTSP 或 token 有问题，直接回报失败。
		if p.cmds[key] == entry {
			p.stopStreamLocked(key)
		}
		if result.err == nil {
			return errors.New("推流进程在启动确认前退出")
		}
		return result.err
	case <-time.After(2 * time.Second):
		// 运行超过两秒认为启动成功，后续异常会通过进程退出日志和服务端超时机制兜底。
		return nil
	}
}

func streamKey(command model.StartCommand) string {
	sourceType := command.SourceType
	if sourceType == "" {
		sourceType = "ROBOT"
	}
	sourceID := command.SourceID
	if sourceID == "" {
		sourceID = command.DeviceID
	}
	if sourceID == "" {
		sourceID = command.RobotID
	}
	roomName := command.RoomName
	if roomName == "" {
		roomName = command.Channel + "|" + command.Quality
	}
	return strings.Join([]string{sourceType, sourceID, roomName}, "|")
}

func trackName(command model.StartCommand) string {
	return "video." + command.Channel + "." + command.Quality
}

func tokenUsable(expiresAt time.Time) bool {
	return !expiresAt.IsZero() && expiresAt.After(time.Now().Add(30*time.Second))
}

func entryRunning(entry *processEntry) bool {
	return entry != nil && entry.cmd != nil && entry.cmd.Process != nil && entry.cmd.ProcessState == nil
}

func (p *ProcessPublisher) bindSessionLocked(sessionID string, key string) {
	if sessionID == "" {
		return
	}
	p.sessions[sessionID] = key
	if p.streamSessions[key] == nil {
		p.streamSessions[key] = make(map[string]struct{})
	}
	p.streamSessions[key][sessionID] = struct{}{}
}

func (p *ProcessPublisher) unbindSessionLocked(sessionID string) string {
	key := p.sessions[sessionID]
	if key == "" {
		return ""
	}
	delete(p.sessions, sessionID)
	delete(p.streamSessions[key], sessionID)
	if len(p.streamSessions[key]) == 0 {
		delete(p.streamSessions, key)
	}
	return key
}

func (p *ProcessPublisher) watchProcessExit(key string, entry *processEntry, sessionID string) {
	go func() {
		result := <-entry.done
		p.mu.Lock()
		defer p.mu.Unlock()
		if p.cmds[key] != entry {
			return
		}
		delete(p.cmds, key)
		sessionIDs := p.sessionIDsLocked(key)
		p.unbindStreamLocked(key)
		p.unexpectedExits++
		p.recordCleanupLocked(result.exitedAt)
		log.Printf("推流进程异常退出并已清理资源，模式=%s 会话ID=%s：%v", entry.mode, sessionID, result.err)
		p.emitLocked(LifecycleEvent{SessionIDs: sessionIDs, ReasonCode: "PUBLISH_PROCESS_EXITED", Message: "推流进程异常退出"})
	}()
}

func (p *ProcessPublisher) watchTokenExpiry(key string, entry *processEntry, sessionID string) {
	delay := time.Until(entry.expiresAt)
	if delay <= 0 {
		delay = time.Millisecond
	}
	go func() {
		timer := time.NewTimer(delay)
		defer timer.Stop()
		<-timer.C
		p.mu.Lock()
		defer p.mu.Unlock()
		if p.cmds[key] != entry {
			return
		}
		p.tokenExpirations++
		log.Printf("发布 Token 到期，开始清理推流资源，会话ID=%s", sessionID)
		p.emitLocked(LifecycleEvent{SessionIDs: p.sessionIDsLocked(key), ReasonCode: "PUBLISH_TOKEN_EXPIRED", Message: "发布 Token 已到期"})
		_ = p.stopStreamLocked(key)
	}()
}

func (p *ProcessPublisher) forceKillAfterTimeout(entry *processEntry) {
	timeout := p.cfg.PublisherStopTimeout
	if timeout <= 0 {
		timeout = 5 * time.Second
	}
	go func() {
		timer := time.NewTimer(timeout)
		defer timer.Stop()
		<-timer.C
		if entry.cmd == nil || entry.cmd.Process == nil || entry.cmd.ProcessState != nil {
			return
		}
		if err := syscall.Kill(-entry.cmd.Process.Pid, syscall.SIGKILL); err != nil {
			_ = entry.cmd.Process.Kill()
		}
		p.mu.Lock()
		p.forcedKills++
		p.mu.Unlock()
		log.Printf("推流进程未在宽限期内退出，已强制清理，模式=%s", entry.mode)
	}()
}

func (p *ProcessPublisher) unbindStreamLocked(key string) {
	for sessionID, sessionKey := range p.sessions {
		if sessionKey == key {
			delete(p.sessions, sessionID)
		}
	}
	delete(p.streamSessions, key)
}

func (p *ProcessPublisher) recordCleanupLocked(exitedAt time.Time) {
	millis := time.Since(exitedAt).Milliseconds()
	if millis < 0 {
		millis = 0
	}
	p.lastCleanupMillis = millis
	if millis > p.maxCleanupMillis {
		p.maxCleanupMillis = millis
	}
}

func (p *ProcessPublisher) Snapshot() Snapshot {
	p.mu.Lock()
	defer p.mu.Unlock()
	return Snapshot{
		ActivePublishers: len(p.cmds), ActiveSessions: len(p.sessions),
		UnexpectedExits: p.unexpectedExits, TokenExpirations: p.tokenExpirations,
		ForcedKills: p.forcedKills, LastCleanupMillis: p.lastCleanupMillis,
		MaxCleanupMillis: p.maxCleanupMillis,
	}
}

func (p *ProcessPublisher) Events() <-chan LifecycleEvent {
	return p.events
}

func (p *ProcessPublisher) sessionIDsLocked(key string) []string {
	result := make([]string, 0, len(p.streamSessions[key]))
	for sessionID := range p.streamSessions[key] {
		result = append(result, sessionID)
	}
	return result
}

func (p *ProcessPublisher) emitLocked(event LifecycleEvent) {
	if len(event.SessionIDs) == 0 {
		return
	}
	select {
	case p.events <- event:
	default:
		log.Printf("推流生命周期事件队列已满，事件未下发，原因码=%s", event.ReasonCode)
	}
}

File: python-client/robot_media_client/publisher.py
"""外部视频推流进程管理模块。

本模块只负责把“已经确认可访问的 RTSP 地址”交给外部进程发布到 LiveKit。
RTSP 地址解析和探测发生在 `mqtt_client.py`，这里不再重复探测。

推流模式优先级：

1. `PUBLISHER_CMD` 非空：完全使用自定义命令，适合现场临时接入特殊推流器。
2. `PUBLISHER_MODE=ffmpeg`：直接使用 `FFMPEG_PUBLISHER_CMD`。
3. `PUBLISHER_MODE=gstreamer`：只使用 `gstreamer-publisher`，失败直接抛错。
4. `PUBLISHER_MODE=auto`：优先 GStreamer，失败或短时间退出时回退 FFmpeg。

所有外部进程都按 sessionId 管理。收到同一个 sessionId 的新 start 时，会先停止
旧进程再启动新进程；收到 stop 或 MQTT 断线时，会杀掉对应进程组，避免后台残留
继续占用摄像头或 LiveKit 房间。
"""

from __future__ import annotations

import os
import shlex
import signal
import subprocess
import threading
import time

from .config import Config
from .model import StartCommand


class ProcessPublisher:
    """用外部进程把 RTSP 视频发布到 LiveKit。

    类本身不直接处理媒体流，只负责：

    - 选择 GStreamer、FFmpeg 或自定义命令。
    - 渲染命令模板中的 RTSP/LiveKit/token 占位符。
    - 启停和观察外部 publisher 子进程。
    - 为上层返回稳定的 trackSid/trackName 占位值。
    """

    def __init__(self, cfg: Config) -> None:
        """保存配置，并按 sessionId 管理 publisher 子进程。

        `gstreamer_failed_rtsp_urls` 是 auto 模式下的本地经验缓存：某个 RTSP URL
        已经触发过 GStreamer 失败后，后续同 URL 直接先走 FFmpeg，减少失败等待。
        """
        self.cfg = cfg
        self.processes: dict[str, subprocess.Popen[bytes]] = {}
        self.gstreamer_failed_rtsp_urls: set[str] = set()
        self.lock = threading.RLock()

    def start(self, command: StartCommand, rtsp_url: str) -> tuple[str, str]:
        """启动指定 session 的推流；同 session 重复 start 会先清理旧进程。

        `command` 来自 MQTT start/switch-channel 指令，里面包含 LiveKit URL、
        publisher token、roomName、channel、quality 等信息。`rtsp_url` 是上层
        已经解析并通过 ffprobe 探测的本地摄像头地址。

        返回值：

        - trackSid：当前外部 publisher 不会把真实 LiveKit track sid 回传给父进程，
          因此客户端用 `TR_{sessionId}` 作为稳定占位值。
        - trackName：按 `video.{channel}.{quality}` 生成，便于后端/前端识别。
        """
        with self.lock:
            # switch-channel 或重复 start 可能复用同一个 sessionId；先停旧进程，保证
            # 同一 session 只有一个 publisher 占用资源。
            self._stop_locked(command.session_id)
            track_name = "video." + command.channel + "." + command.quality
            if self.cfg.publisher_cmd:
                # PUBLISHER_CMD 是最高优先级逃生口，适合现场验证新 pipeline。
                return self._start_command(command, rtsp_url, track_name, self.cfg.publisher_cmd, "custom")
            if self.cfg.publisher_mode == "ffmpeg":
                if not self.cfg.ffmpeg_publisher_cmd:
                    raise RuntimeError("PUBLISHER_MODE=ffmpeg requires FFMPEG_PUBLISHER_CMD")
                # 强制 ffmpeg 模式通常用于 GStreamer 在目标硬件或 RTSP 源上不稳定的场景。
                return self._start_command(command, rtsp_url, track_name, self.cfg.ffmpeg_publisher_cmd, "ffmpeg")
            if self.cfg.publisher_mode not in ("auto", "gstreamer"):
                raise RuntimeError(f"unsupported PUBLISHER_MODE: {self.cfg.publisher_mode}")
            if self._should_start_with_ffmpeg(command, rtsp_url):
                # auto 模式下的“先用 FFmpeg”分支，来自人工配置或历史失败缓存。
                return self._start_command(command, rtsp_url, track_name, self.cfg.ffmpeg_publisher_cmd, "ffmpeg-first")
            try:
                result = self._start_gstreamer(command, rtsp_url, track_name)
                if self.cfg.publisher_mode == "auto" and self.cfg.ffmpeg_publisher_cmd:
                    # 有些 RTSP 流能让进程启动成功，但会在几秒内因硬件编解码或 pipeline 问题退出。
                    self._watch_gstreamer_for_fallback(command, rtsp_url, track_name)
                return result
            except Exception:
                if self.cfg.publisher_mode == "gstreamer" or not self.cfg.ffmpeg_publisher_cmd:
                    raise
                # GStreamer 启动阶段失败，auto 模式立即降级 FFmpeg，并记住这个 RTSP URL。
                self.gstreamer_failed_rtsp_urls.add(rtsp_url)
                print("publisher fallback ffmpeg", command.session_id, flush=True)
                return self._start_command(command, rtsp_url, track_name, self.cfg.ffmpeg_publisher_cmd, "ffmpeg")

    def _start_command(
        self,
        command: StartCommand,
        rtsp_url: str,
        track_name: str,
        template: str,
        mode: str,
    ) -> tuple[str, str]:
        """按模板渲染自定义 publisher 命令并启动进程。

        这里同时服务 `PUBLISHER_CMD` 和 `FFMPEG_PUBLISHER_CMD`。模板会通过
        `shlex.split()` 拆成 argv，避免 shell 拼接带来的注入问题；如果确实需要
        shell 语法，应把复杂逻辑放到脚本里，再让模板调用脚本。
        """
        args = self._render_args(template, command, rtsp_url, track_name)
        if not args:
            raise RuntimeError("publisher command is empty")
        # start_new_session=True 会让子进程成为新进程组组长，后续 stop 时可以杀整个
        # 进程组，包括脚本内部再拉起的 ffmpeg/gstreamer-publisher 子进程。
        process = subprocess.Popen(args, stdout=None, stderr=None, start_new_session=True)
        print("publisher command", mode, " ".join(shlex.quote(arg) for arg in args), flush=True)
        self.processes[command.session_id] = process
        self._ensure_running(command.session_id, process)
        return "TR_" + command.session_id, track_name

    def _should_start_with_ffmpeg(self, command: StartCommand, rtsp_url: str) -> bool:
        """在 auto 模式下，对人工指定或已知 GStreamer 失败过的流直接使用 ffmpeg。

        两类流会跳过 GStreamer：

        - RTSP URL 在本进程生命周期内已经触发过 GStreamer 失败。
        - deviceId 被配置在 `PUBLISHER_FFMPEG_FIRST_DEVICE_IDS`。
        """
        if self.cfg.publisher_mode != "auto" or not self.cfg.ffmpeg_publisher_cmd:
            return False
        if rtsp_url in self.gstreamer_failed_rtsp_urls:
            return True
        return command.device_id in self.cfg.publisher_ffmpeg_first_device_ids

    def _watch_gstreamer_for_fallback(self, command: StartCommand, rtsp_url: str, track_name: str) -> None:
        """在 auto 模式下短暂观察 GStreamer 子进程，崩溃时自动切到 ffmpeg。

        `_ensure_running()` 只观察启动后的前 2 秒；有些 pipeline 会在 3-8 秒内
        因 RTSP payload、硬件解码或 EGL/display 问题退出，所以 auto 模式额外开
        一个 daemon 线程观察 `PUBLISHER_FALLBACK_WATCH_SECONDS`。
        """
        process = self.processes.get(command.session_id)
        if process is None:
            return
        thread = threading.Thread(
            target=self._fallback_if_gstreamer_exits,
            args=(command, rtsp_url, track_name, process),
            name=f"publisher-fallback-{command.session_id}",
            daemon=True,
        )
        thread.start()

    def _fallback_if_gstreamer_exits(
        self,
        command: StartCommand,
        rtsp_url: str,
        track_name: str,
        process: subprocess.Popen[bytes],
    ) -> None:
        """如果 GStreamer 直推在观察窗口内退出，且 session 仍有效，就改用 ffmpeg 推同一路流。

        这里会再次检查 `self.processes[sessionId] is process`，防止以下竞态：

        - 用户已经 stop 了该 session。
        - 后端又下发了同 session 的新 start。
        - MQTT 断线触发了 stop_all。

        只有当前仍由这个 GStreamer 进程负责的 session，才允许自动 fallback。
        """
        deadline = time.monotonic() + self.cfg.publisher_fallback_watch_seconds
        while time.monotonic() < deadline:
            code = process.poll()
            if code is None:
                time.sleep(0.2)
                continue
            with self.lock:
                if self.processes.get(command.session_id) is not process:
                    return
                self.processes.pop(command.session_id, None)
                self.gstreamer_failed_rtsp_urls.add(rtsp_url)
                print(
                    "publisher auto fallback ffmpeg",
                    command.session_id,
                    f"gstreamer_exit={code}",
                    flush=True,
                )
                try:
                    self._start_command(command, rtsp_url, track_name, self.cfg.ffmpeg_publisher_cmd, "ffmpeg")
                except Exception as exc:
                    print("publisher auto fallback failed", command.session_id, exc, flush=True)
            return

    def _start_gstreamer(self, command: StartCommand, rtsp_url: str, track_name: str) -> tuple[str, str]:
        """使用 gstreamer-publisher 默认路径发布 RTSP 到 LiveKit。

        `gstreamer-publisher` 负责连接 LiveKit 并发布 track；`GSTREAMER_PIPELINE`
        只描述 `--` 后面的媒体处理 pipeline。默认 pipeline 读取 RTSP H264，
        depay 后交给 h264parse，再由 publisher 推到 LiveKit。
        """
        pipeline = self._render_text(self.cfg.gstreamer_pipeline, command, rtsp_url, track_name)
        args = [
            self.cfg.gstreamer_publisher_path,
            "--url",
            command.livekit_url,
            "--token",
            command.publisher_token,
            "--",
        ]
        args.extend(shlex.split(pipeline))
        process = subprocess.Popen(args, stdout=None, stderr=None, start_new_session=True)
        print("publisher command", " ".join(shlex.quote(arg) for arg in args), flush=True)
        self.processes[command.session_id] = process
        self._ensure_running(command.session_id, process)
        return "TR_" + command.session_id, track_name

    def stop(self, session_id: str) -> None:
        """停止指定 session 的 publisher 进程。

        stop 通常来自 MQTT `media/video/stop`，也可能来自同 session 的重新 start。
        """
        with self.lock:
            self._stop_locked(session_id)

    def stop_all(self) -> None:
        """停止当前客户端管理的所有 publisher 进程。

        MQTT 断线或客户端退出时调用，保证机器人端不会继续向旧 LiveKit room 推流。
        """
        with self.lock:
            for session_id in list(self.processes.keys()):
                self._stop_locked(session_id)

    def _stop_locked(self, session_id: str) -> None:
        """在已持锁的情况下停止并移除一个 publisher 进程。

        优先 kill 进程组，确保脚本拉起的子进程一并结束；如果进程组 kill 失败，
        再退回 kill 主进程。
        """
        process = self.processes.pop(session_id, None)
        if process is None:
            return
        if process.poll() is not None:
            return
        try:
            os.killpg(process.pid, signal.SIGKILL)
        except OSError:
            process.kill()

    def _ensure_running(self, session_id: str, process: subprocess.Popen[bytes]) -> None:
        """短暂观察子进程是否立即退出，尽早暴露 pipeline 参数错误。

        外部 publisher 如果 token、LiveKit URL、RTSP pipeline 参数错误，通常会很快
        退出。这里等待 2 秒，如果还活着就认为“启动成功”，后续运行时异常由进程日志、
        fallback 观察线程或后端超时机制兜底。
        """
        deadline = time.monotonic() + 2
        while time.monotonic() < deadline:
            code = process.poll()
            if code is not None:
                self.processes.pop(session_id, None)
                raise RuntimeError(f"publisher exited with code {code}")
            time.sleep(0.05)

    def _render_args(self, template: str, command: StartCommand, rtsp_url: str, track_name: str) -> list[str]:
        """把命令模板拆分为 argv，并替换会话、RTSP 和 LiveKit 占位符。

        支持的占位符：

        - `{rtsp}`：最终使用的 RTSP URL。
        - `{livekitUrl}`：LiveKit 服务地址。
        - `{token}`：publisher token。
        - `{room}`：LiveKit room 名称。
        - `{track}`：客户端生成的 trackName。
        """
        return [
            self._render_text(part, command, rtsp_url, track_name)
            for part in shlex.split(template)
        ]

    @staticmethod
    def _render_text(template: str, command: StartCommand, rtsp_url: str, track_name: str) -> str:
        """替换单个模板片段中的占位符。"""
        return (
            template
            .replace("{rtsp}", rtsp_url)
            .replace("{livekitUrl}", command.livekit_url)
            .replace("{token}", command.publisher_token)
            .replace("{room}", command.room_name)
            .replace("{track}", track_name)
        )
